packages/rum-core/src/domain/rumEventsCollection/action/rageClickChain.ts

```diff
@@ -0,0 +1,123 @@
+import { monitor, ONE_SECOND, timeStampNow } from '@datadog/browser-core'
+import { FrustrationType } from '../../../rawRumEvent.types'
+import type { Click } from './trackClickActions'
+
+export interface RageClickChain {
+  tryAppend: (click: Click) => boolean
+  stop: () => void
+}
+
+export const MAX_DURATION_BETWEEN_CLICKS = ONE_SECOND
+export const MAX_DISTANCE_BETWEEN_CLICKS = 100
+
+const enum RageClickChainStatus {
+  WaitingForMoreClicks,
+  WaitingForClicksToStop,
+  Finalized,
+}
+
+export function createRageClickChain(firstClick: Click): RageClickChain {
+  const bufferedClicks: Click[] = []
+  let stoppedClicksCount = 0
+  let status = RageClickChainStatus.WaitingForMoreClicks
+  let timeout: number | undefined
+  const rageClick = firstClick.clone()
+
+  function dontAcceptMoreClick() {
+    if (status === RageClickChainStatus.WaitingForMoreClicks) {
+      status = RageClickChainStatus.WaitingForClicksToStop
+      tryFinalize()
+    }
+  }
+
+  function tryFinalize() {
+    if (status === RageClickChainStatus.WaitingForClicksToStop && stoppedClicksCount === bufferedClicks.length) {
+      status = RageClickChainStatus.Finalized
+      finalizeClicks(bufferedClicks, rageClick)
+    }
+  }
+
+  function appendClick(click: Click) {
+    click.onStop(() => {
+      stoppedClicksCount += 1
+      tryFinalize()
+    })
+    bufferedClicks.push(click)
+    timeout = setTimeout(monitor(dontAcceptMoreClick), MAX_DURATION_BETWEEN_CLICKS)
+  }
+
+  appendClick(firstClick)
+  return {
+    tryAppend: (click) => {
+      clearTimeout(timeout)
+
+      if (status !== RageClickChainStatus.WaitingForMoreClicks) {
+        return false
+      }
+
+      if (
+        bufferedClicks.length > 0 &&
+        !areEventsSimilar(bufferedClicks[bufferedClicks.length - 1].event, click.event)
+      ) {
+        dontAcceptMoreClick()
+        return false
+      }
+
+      appendClick(click)
+      return true
+    },
+    stop: () => {
+      clearTimeout(timeout)
+      dontAcceptMoreClick()
+    },
+  }
+}
+
+/**
+ * Checks whether two events are similar by comparing their target, position and timestamp
+ */
+function areEventsSimilar(first: MouseEvent, second: MouseEvent) {
+  return (
+    first.target === second.target &&
+    mouseEventDistance(first, second) <= MAX_DISTANCE_BETWEEN_CLICKS &&
+    first.timeStamp - second.timeStamp <= MAX_DURATION_BETWEEN_CLICKS
+  )
+}
+
+function mouseEventDistance(origin: MouseEvent, other: MouseEvent) {
+  return Math.sqrt(Math.pow(origin.clientX - other.clientX, 2) + Math.pow(origin.clientY - other.clientY, 2))
+}
+
+function finalizeClicks(clicks: Click[], rageClick: Click) {
+  if (isRage(clicks)) {
+    // If it should be be considered as a rage click, discard individual clicks and
+    // validate the rage click.
+    clicks.forEach((click) => {
+      click.discard()
+      click.getFrustrations().forEach((frustration) => {
+        rageClick.addFrustration(frustration)
+      })
```

Comment:
If `addFrustration` could also accept an array of frustrations we could remove this nested foreach

Reply:
The nested forEach would still be there but spread into another function, right? Are you suggesting this change only for clarity?

Reply:
Yes, I'm in favor of moving the iteration to addFrustration to simplify the code here

Reply:
After discussion, we chose to keep it like this because:
* frustrations are in a `Set` here, and having `addFrustration(frustration: FrustrationType | Set<FrustrationType>)` is less idiomatic than `addFrustration(frustration: FrustrationType | FrustrationType[])`.
* moving this to the `Click` implementation means that we need to support it in the mocked version used in tests.

While not big drawbacks, we decided that it wasn't worth it and to leave it like.